go/pkg/common/utils.go

```diff
@@ -124,3 +125,52 @@ func SyncMapLength(m *sync.Map) int {
 	})
 	return i
 }
+
+// RetryUntilSuccessful calls callback every interval until duration until it returns false (to not retry)
+func RetryUntilSuccessful(duration time.Duration,
+	interval time.Duration,
+	callback func() (bool, error)) error {
+
+	wrapFunctionNoResult := func() (interface{}, bool, error) {
+		retry, err := callback()
+		return nil, retry, err
+	}
+
+	_, err := RetryUntilSuccessfulWithResult(duration, interval, wrapFunctionNoResult)
+	return err
+}
+
+// RetryUntilSuccessfulWithResult calls callback every interval until duration until it returns false (to not retry),
+// and returns the result of the callback
+func RetryUntilSuccessfulWithResult(duration time.Duration,
+	interval time.Duration,
+	callback func() (interface{}, bool, error)) (interface{}, error) {
+	var (
+		lastErr, err error
+		result       interface{}
+		shouldRetry  bool
+	)
+
+	timedOutErrorMessage := "Timed out waiting until successful"
+	deadline := time.Now().Add(duration)
+
+	// while we haven't passed the deadline
+	for !time.Now().After(deadline) {
+		result, shouldRetry, err = callback()
+		lastErr = err
+		if !shouldRetry {
+			return result, err
+		}
+		time.Sleep(interval)
+		continue
+
+	}
+	if lastErr != nil {
+
+		// wrap last error
+		return result, errors.Wrapf(lastErr, timedOutErrorMessage)
+	}
+
+	// duration expired without any last error
+	return result, errors.Errorf(timedOutErrorMessage)
```

Comment:
it stem of a bad design, returning an error though last attempt returned no error.  it is fair that deadline exceeded and request suceeded. perhaps, a glitch on timing, why exploding then